source/module_io/json_output/abacusjson.h

```diff
@@ -54,7 +54,33 @@ using jsonValue = rapidjson::Value;
 class AbacusJson
 {
   public:
-    // The template specialization method adds a key-val object to the doc tree
+    /**
+     *  @brief: The template specialization method adds value to the doc tree
+
+     *  @param: 'keys' is a vector string, represents the path to be added to the json tree.
+     *
+     *          'value' is the value that needs to be added to the json tree, which type can be 
+     *          Json::jsonValue or other common value type(such as int, double ,bool ,std::string).
+     *
+     *          'IsArray' is a bool value, means the whether the root node to which 'value' is added is an array.
+     *
+     *  @usage: 1. Add/Modify a double val to object json node (key2 is a object node): 
+     *                Json::AbacusJson::add_json({"key1","key2"}, 3.1415,false);
+     *
+     *          2. Pushback a double val to array json node (key2 is a array node):
+     *                Json::AbacusJson::add_json({"key1","key2"}, 3.1415,true);
+     *
+     *          3. Modify a doble val to array json node (key2 is a array node), when use this method, 
+     *            The index number of the array starts at 1, and if the index is 0, it means that the last element 
```

Comment:
Here, perhaps we can follow the indexing method of a Python list, where numbers>=0 start from the first and numbers <0 start from the last: for example, 0 represents the first and -1 represents the last?

Reply:
Is this situation more suitable for writing an overloaded function, passing the idex that needs to be modified as an int parameter?

Reply:
Considering that there may be an array whose key is a number string?

Reply:
> Here, perhaps we can follow the indexing method of a Python list, where numbers>=0 start from the first and numbers <0 start from the last: for example, 0 represents the first and -1 represents the last?

Good idea, I will change this function as this.

Reply:
> Is this situation more suitable for writing an overloaded function, passing the idex that needs to be modified as an int parameter?

I think this is not a convenient way to implement and use the design. For the current function, the required key argument is a string array. 

It also works correctly if the json path has multiple arrays.  For an example, we can use this to add a multi-array-path
```
 Json::AbacusJson::add_json({"key1","1","key2","key3","1"}, 3.1415,true); // (key1,key3 both array)
```
 If the array subscripts are separated, another subscripted array is required, the function becomes more complex to use, and the user input is less friendly.

Reply:
> Considering that there may be an array whose key is a number string?

I think that since rapidjson is used in abacus, developers need to avoid this purely numeric json key situation. 
According to the json code specification: https://pantao.gitbooks.io/coding-standards/content/json.html

Json property attribute name  specifications are as follows:

` ` `
The attribute name should be a meaningful name with defined semantics.
The attribute name must be a humped, ASCII character string.
The first character must be a letter, underscore (_), or dollar sign ($).
Other characters that follow can be letters, numbers, underscores (_), or dollar signs ($).
You should avoid using reserved keywords in Javascript (see below for a list of reserved Javascript words)
` ` `

Reply:
> > Considering that there may be an array whose key is a number string?
> 
> I think that since rapidjson is used in abacus, developers need to avoid this purely numeric json key situation. According to the json code specification: https://pantao.gitbooks.io/coding-standards/content/json.html
> 
> Json property attribute name specifications are as follows:
> 
> ```
> The attribute name should be a meaningful name with defined semantics.
> 
> The attribute name must be a humped, ASCII character string.
> 
> The first character must be a letter, underscore (_), or dollar sign ($).
> 
> Other characters that follow can be letters, numbers, underscores (_), or dollar signs ($).
> 
> You should avoid using reserved keywords in Javascript (see below for a list of reserved Javascript words)
> ```
It should be noted that this is a specification, which means that using numeric strings as keys is valid for JSON. We should not ignore this situation when design the function.

Reply:
> It should be noted that this is a specification, which means that using numeric strings as keys is valid for JSON. We should not ignore this situation when design the function.

I think since this is the json tool used internally by abacus, maybe developers should follow the json writing recommendations and develop a canonical json format. 

By the way I will consider your suggestion and see if there is a better way!

Reply:
> It should be noted that this is a specification, which means that using numeric strings as keys is valid for JSON. We should not ignore this situation when design the function.

Now support these feature!
1. Array start by 0. Same as python. If the number is negative, the array traverses from the back, with -1 representing the last one.
2. Purely numeric objects are supported. Here is an example：
```
    Json::AbacusJson::add_json({"0","11"}, true,true);
    Json::AbacusJson::add_json({"0","11"}, false,true);
    Json::AbacusJson::add_json({"0","11","-1"}, 1,true);


=>json:
    "0": {
        "11": [
            true,
            1
        ]
    }
```